### PR TITLE
WIP Improve NUTS tuning

### PR DESCRIPTION
**Pull request by twiecki**
I started experimenting with better NUTS initialization. What stan does is to run a NUTS chain and then estimate the posterior covariance from that run. This cov matrix is used as the mass matrix for the actual NUTS run.

This PR removes some options of initialization (like MAP) which I think are useless but adds a single path. We also use LedoitWolf regularization here like STAN does. I get better sampling for the tricky hierarchical GLM model.

This PR also makes obvious that NUTS' hyperparameters can't be properly set, as internally we only compute the integration times in log-space, but allow passing in non-log, so this setting of NUTS parameters externally is a bit clunky. 

Suggestions welcome.

### CODE REVIEW

**Comment by twiecki** [comment]
Without NUTS init:
image

With NUTS init:
image

Still problematic, might require RMHMC.

File: pymc3/sampling.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def assign_step_methods(model, step=None, methods=(NUTS, HamiltonianMC, Metropol
     return steps
 
 
-def sample(draws, step=None, init='advi', n_init=200000, start=None,
+def sample(draws, step=None, n_advi=200000, n_nuts=2000, start=None,
```
Comment (fonnesbeck) [inline pymc3/sampling.py:84]: I'm a little uncomfortable encoding particular algorithm names in the arguments of `sample`. Can we just not condition internally depending on which step is used?

Comment (twiecki) [reply pymc3/sampling.py:84]: Good point. I'm starting to think about this as init-programs. This new one uses ADVI+NUTS. Perhaps just an init_kwargs dict that can get passed is most general?

Comment (fonnesbeck) [reply pymc3/sampling.py:84]: I think so. It limits what we can provide in terms of a docstring, but keeps things general. Another option would be a more explicit `init_params` argument that could vary depending on the init method chosen.

Comment (twiecki) [reply pymc3/sampling.py:84]: Yep exactly. There could be different init-functions and we always forward **init_params.

**Comment by junpenglao** [comment]
@twiecki I think this can be close now?

**Comment by junpenglao** [comment]
close in favor of https://github.com/pymc-devs/pymc3/pull/2327

**Pull request closed without merging**